### ISSUE TITLE
client-side routing link interception with base url

### ISSUE THREAD

**Issue by yxonic** [brillout/vite-plugin-ssr#205]
If I have `<a href="/bar">` and base url set as `/foo/`, the link would set url to `/bar` instead of `/foo/bar`. However, when I have `<a href="/foo/bar">`, the link would be correct but the page component fail to resolve, and fallback to server-side routing. I wonder if there is a bug, or should I manually implement a link component myself?

**Comment by yxonic** [comment]
I also tried relative path `<a href="./bar">` to avoid hard-coding base url, but got an assertion error:

```
prefetch.ts:76 Uncaught (in promise) Error: [vite-plugin-ssr@0.3.27][Internal Failure] You stumbled upon a bug in `vite-plugin-ssr`'s source code (an internal `assert()` failed). This should definitely not be happening, and you should create a new GitHub issue at https://github.com/brillout/vite-plugin-ssr/issues/new that includes this error stack (the error stack is usually enough to debug internal errors). Or reach out on Discord. A fix will be written promptly. Debug info (this is for the `vite-plugin-ssr` maintainers; you can ignore this): `{"url":"./demos/project"}`.
    at parseWithNewUrl (parseUrl.ts:126)
    at getUrlPathname (parseUrl.ts:52)
    at getPrefetchUrl (prefetch.ts:88)
    at isAlreadyPrefetched (prefetch.ts:80)
    at prefetch.ts:54
    at Array.forEach (<anonymous>)
    at addLinkPrefetchHandlers (prefetch.ts:45)
    at useClientRouter.ts:127
    at async fetchAndRender (useClientRouter.ts:129)
```

**Comment by brillout** [comment]
Yes see https://vite-plugin-ssr.com/base-url

Leaving this open until the error message is fixed.

**Comment by yxonic** [comment]
An update: I got the following error after setting base url as `/foo/` and link to `<a href="/foo/bar">`, before page redirects. I have setup base parameter for both vite and createPageRenderer. Did I miss something?
```
useClientRouter.ts:142 Uncaught (in promise) Error: [vite-plugin-ssr@0.3.27][Internal Failure] You stumbled upon a bug in `vite-plugin-ssr`'s source code (an internal `assert()` failed). This should definitely not be happening, and you should create a new GitHub issue at https://github.com/brillout/vite-plugin-ssr/issues/new that includes this error stack (the error stack is usually enough to debug internal errors). Or reach out on Discord. A fix will be written promptly.
    at getPageContextFromRoute (getPageContext.ts:116)
    at async getPageContext (getPageContext.ts:51)
    at async fetchAndRender (useClientRouter.ts:104)
```

**Comment by brillout** [comment]
I will have a look

**Comment by brillout** [comment]
Improvements released in `0.3.28`.

Let me know if you still run into problems.

**Comment by yxonic** [comment]
Thanks for your effort! The error message is now clearer, but I think the problem of falling back to server side routing still exists.

> If I have `<a href="/bar">` and base url set as `/foo/`, the link would set url to `/bar` instead of `/foo/bar`. However, when I have `<a href="/foo/bar">`, the link would be correct but the page component fail to resolve, and fallback to server-side routing.

After a bit of thinking, I believe the client router should have a base url parameter. This should solve this problem, remove the need of a custom component (now I can just always use `/bar` in href as if there wasn't a base url, and `useClientRouter` will prepend `/foo/` for me), and more elegant (we don't need to think about base urls after client side initialization/hydration, and current resolving logic can remain untouched).

A small issue is that when we **do** want to route to `/bar` instead of `/foo/bar`, the client router wouldn't know. I think this can be resolved by writing `<a href="/bar" rel="external">` to completely bypass client-side routing, which is basically what it should be doing.

Yet another more conservative solution is to not modify url: we still write `<a href="/foo/bar">` (in our custom component), the client router jumps to the exact location, while resolving page `bar` instead of `/foo/bar`. Nonetheless, client router should know about base url anyway, or else it wouldn't know whether to fetch `/foo/bar` or just `bar`. And it also means that the resolving logic should change, which requires more work.

Let me know which design you think is the best, and how long do you expect this to finish. I would like to contribute, but may not have time to work on it until next week. BTW, I love how your library works and how clear the design is. I always believe hiding but not abstracting details introduce more trouble than convenience, and your library have done it better than many.

**Comment by brillout** [comment]
If the user right clicks on `<a href="/bar" />` and opens the link in a new browser tab, then `vite-plugin-ssr` cannot (should not) intercept that. This means that the Base URL needs to be included in all `<a>` links.

Relative links are brittle and (rightfully) not commonly used.

I still don't see a design flaw or a way to improve the current implementation. Feel free to further object.

I'm glad you like the plugin :-)

**Comment by yxonic** [comment]
I think you are right about this. Maybe I will consider writing a vite plugin to rewrite all hrefs.

However, the resolving problem still exists after updating to `0.3.28`:

```
Uncaught (in promise) Error: [vite-plugin-ssr@0.3.28][Wrong Usage][404] Page /foo/bar does not exist. (`vite-plugin-ssr` will now server-side route to `/foo/bar`.)
    at getPageContextFromRoute (getPageContext.ts:118)
    at async getPageContext (getPageContext.ts:51)
    at async fetchAndRender (useClientRouter.ts:104)
```

**Comment by brillout** [comment]
> I think you are right about this. Maybe I will consider writing a vite plugin to rewrite all hrefs.

👍 Keep me updated.

> However, the resolving problem still exists after updating to `0.3.28`:

Are you sure you properly set the `base` option in both `createPageRenderer({ base })` and `vite.config.js`?

See https://github.com/brillout/vite-plugin-ssr/tree/master/examples/base-url

**Comment by yxonic** [comment]
> Are you sure you properly set the `base` option in both `createPageRenderer({ base })` and `vite.config.js`?
>
> See https://github.com/brillout/vite-plugin-ssr/tree/master/examples/base-url

I made a minimal example following your link to demonstrate the problem: https://github.com/yxonic/base-url-with-client-routing

Run the following, and click the links, then error will occur:
```
npm i && npm run prod:server
```

Changes I've made: I changed to client-side routing following your documentation, and added a hook before page unload to let the error message stay.

**Comment by brillout** [comment]
I could reproduce. Let me fix it.

**Comment by brillout** [comment]
I see what's going on, I'll fix this later today.

**Comment by brillout** [comment]
Released in `v0.3.29`.
